### ISSUE TITLE
Building app using v4.0.0 fails with LDSwiftEventSource/LDSwiftEventSource.modulemap not found

### ISSUE THREAD

**Issue by fnando** [launchdarkly/react-native-client-sdk#68]
**Describe the bug**

After upgrading to v4.0.0, I get an error whenever I try to build the app using `Release` mode.

Then I ran the following commands:

```
$ cd ios
$ pod cache clean --all
$ pod install
```

Following this, I started to install the app using Release mode.

```
$ react-native run-ios --scheme=vibrant-testnet --configuration=Release --simulator 'iPhone SE (2nd generation)'
```

**To reproduce**

Build fails with several outputs like `fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found`.

**Expected behavior**

Builds successfully without errors.

**Logs**

```
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
1 error generated.
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
1 error generated.
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
1 error generated.
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
1 error generated.
3 errors generated.
6 errors generated.
fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
1 error generated.

….

fatal error: module map file '/Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Products/Release-iphonesimulator/LDSwiftEventSource/LDSwiftEventSource.modulemap' not found
1 error generated.


** BUILD FAILED **


The following build commands failed:
	CompileC /Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Intermediates.noindex/vibrant.build/Release-iphonesimulator/vibrant-testnet.build/Objects-normal/x86_64/Vibrant\ Testnet_vers.o /Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Intermediates.noindex/vibrant.build/Release-iphonesimulator/vibrant-testnet.build/DerivedSources/Vibrant\ Testnet_vers.c normal x86_64 c com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Intermediates.noindex/vibrant.build/Release-iphonesimulator/vibrant-testnet.build/Objects-normal/i386/Vibrant\ Testnet_vers.o /Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Intermediates.noindex/vibrant.build/Release-iphonesimulator/vibrant-testnet.build/DerivedSources/Vibrant\ Testnet_vers.c normal i386 c com.apple.compilers.llvm.clang.1_0.compiler
	CompileC /Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Intermediates.noindex/vibrant.build/Release-iphonesimulator/vibrant-testnet.build/Objects-normal/arm64/Vibrant\ Testnet_vers.o /Users/fnando/Library/Developer/Xcode/DerivedData/vibrant-bunfharcsfqcsyhkzymjoyghvrxm/Build/Intermediates.noindex/vibrant.build/Release-iphonesimulator/vibrant-testnet.build/DerivedSources/Vibrant\ Testnet_vers.c normal arm64 c com.apple.compilers.llvm.clang.1_0.compiler
(3 failures)
```

**SDK version**

v4.0.0

**Language version, developer tools**

- react-native 0.63.3
- Xcode 12.4 (12D4e)
- node v12.19.0

**OS/platform**

- macOS 11.12.3 (Mac Mini M1 running with terminal running in x86_64 mode)
- Simulator Version 12.4 (940.20), SimulatorKit 597.13.0.1, CoreSimulator 732.18.6

**Comment by gwhelanLD** [comment]
Hi @fnando,

I'm not able to replicate this issue on a test application, would you be able to provide your `Podfile` either here, or to our [support team](https://support.launchdarkly.com/hc/en-us) so we have the best chance of assisting you with this issue?

Thanks,
@gwhelanLD

**Comment by torchhound** [comment]
Hi @fnando thanks for reporting this. Does your build work in Debug mode? Is this error only on Release?

**Comment by fnando** [comment]
Hey all!

Yes, the app was built successfully with `react-native run-ios --scheme=vibrant-testnet --configuration=Debug --simulator 'iPhone SE (2nd generation)'`, so I'd say it's only happening with `Release` mode.

Here's my Podfile:

```ruby
# frozen_string_literal: true

require_relative "../node_modules/react-native/scripts/react_native_pods"
require_relative "../node_modules/@react-native-community/cli-platform-ios/native_modules"

# react-native-permission' base path.
permissions_path = "../node_modules/react-native-permissions/ios"

# The following method is needed to ensure the "archive" step works in XCode.
# It removes pods that are already included in the main project. Without
# this, you'd see errors when you archive like:
#
# "Multiple commands produce ... libReact.a"
# "Multiple commands produce ... libyoga.a"
def ignore_pods(installer)
  targets_to_ignore = %w[React yoga]
  installer.pods_project.targets.each do |target|
    target.remove_from_project if targets_to_ignore.include? target.name
  end
end

platform :ios, "10.0"

target "vibrant" do
  native_modules = use_native_modules!

  use_react_native!(path: native_modules["reactNativePath"])

  pod "Permission-Camera", path: "#{permissions_path}/Camera.podspec"
  pod "Singular-React-Native",
      podspec: "../node_modules/singular-react-native/Singular-React-Native.podspec"

  target "vibrantTests" do
    inherit! :complete
    # Pods for testing
  end

  # Run post install instructions.
  post_install do |installer|
    ignore_pods(installer)

    installer.pods_project.targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings["IPHONEOS_DEPLOYMENT_TARGET"] = "10.0"
        config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
      end
    end
  end

  target "vibrant-testnet"
  target "vibrant-rc"
end
```

**Comment by gwhelanLD** [comment]
I'm still having some difficulty replicating this. I'm leaning towards suspecting something related to Apple Silicon, which we unfortunately don't have as much testing ability for (🤞 that CircleCI eventually gets Apple Silicon environments). 

A few things I might try on an M1 machine (will be a bit before I can test it myself).

I see that `config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"` seems to be a workaround on Apple Silicon, but some posts seem to indicate `config.build_settings["ONLY_ACTIVE_ARCH"] = "YES"` might be an alternative/addition to try. That may not be an acceptable workaround for app submission however...

`LDSwiftEventSource` does some rather complicated `VALID_ARCHS` and platform configuration. That was written in the early days of Xcode 12. It sounds like Apple is moving away from that method, so we may need to revisit that now. I don't think we've had the opportunity to test building the iOS SDK that the RN SDK uses on an actual Apple Silicon device (though we're actively looking into it). If you're feeling particularly generous, you could try to build our [example iOS app](https://github.com/launchdarkly/hello-ios-swift) to validate whether this is definitely a React Native specific problem.

The settings:
```
config.build_settings["IPHONEOS_DEPLOYMENT_TARGET"] = "10.0"
config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
```
may also be redundant for `LDSwiftEventSource`, so you could also try avoiding those overrides in case it makes a difference. Something like:
```
if target.product_name == "LDSwiftEventSource"then
  next
end
```
before `target.build_configurations.each do |config|`.

One final thing to try that might help is the Cocoapods specifiers discussed [here](https://stackoverflow.com/questions/52910973/how-to-remove-use-frameworks-and-keep-using-swift-pods-on-an-objective-c-projec). 

Sorry I can't give you a more concrete answer at this time. We'll work on getting an Apple Silicon environment on our end to try to replicate in a more similar environment to yours. But I hope some of this may be helpful to finding a workaround in the mean time.

Thanks,
@gwhelanLD

**Comment by todorone** [comment]
@gwhelanLD Update:
- Issue is with the new Apple Silicon. The same project works on x86 and fails to build on M1.
- Every other RN dependency works flawlessly(almost), while LD fails, so the issue is definitely somewhere in iOS config.

Could you please revisit the config(`VALID_ARCHS` is deprecated and doesn't work on XCode 13 afaik) and see if it can be adjusted to support Apple Silicon?
Please let me know if any further info is needed.

**Comment by todorone** [comment]
With rocket fast adoption of M1 devices, is there any update on this? @gwhelanLD

**Comment by gakinson** [comment]
I am also experiencing this issue.

**Comment by todorone** [comment]
@louis-launchdarkly @gwhelanLD This is a kinda blocker for a company to use LD, is there something that could be done with it?

**Comment by Hurtak** [comment]
Some more additional context

- When running development build to simulator, the build succeeds
- When running production build to simulator, the build fails 
- When running production build to simulator with LaunchDarkly removed from the project, the build succeeds 
- When running production build to a real Apple TV device, the build succeeds

So the problem is only for production build to the simulator for some reason. 

Temporary workaround: add  `EXCLUDED_ARCHS = arm64;` to the `ios/rntv.xcodeproj/project.pbxproj` for the production build - that way I was able to run the production build in the simulator. But be aware that this value will probably break you archive action since the real TvOS device runs only on arm64.

**Comment by louis-launchdarkly** [comment]
Thank you @Hurtak for the temporary workaround. We are aware of the issue and it is on our backlog.

Filed Internally as 136671.

**Comment by gwhelanLD** [comment]
Hi everyone,

We believe this issue was resolved in [the 5.1.1 release](https://github.com/launchdarkly/react-native-client-sdk/releases/tag/5.1.1) where we removed the `VALID_ARCHS` workaround from the transient dependency [`LDSwiftEventSource`](https://github.com/launchdarkly/swift-eventsource). With the current `launchdarkly-react-native-client-sdk` I am successfully building the `ManualTestApp` in the repository with `npx react-native run-ios --configuration=Release`. Before I go ahead and close this issue, would anyone be able to verify this fix for their application? @fnando @todorone @Hurtak @gakinson 

Thanks,
@gwhelanLD
